src/test/java/com/endava/internship/collections/StudentSetTest.java

```diff
@@ -0,0 +1,217 @@
+        Iterator it = set.iterator();
+        while(it.hasNext()) {
+            assertNotNull(it.next());
+        }
+    }
+
+    @Test
+    void elementsFromOtherCollectionShouldBeAddedInExactlyOrder(){
+        List<Student> list = Arrays.asList(st1, null, st2);
+
+        assertTrue(set.addAll(list));
+        assertThat(set).containsExactlyInAnyOrder(st1, null, st2);
+    }
+
+    @Test
+    void mustRemoveAllElementsContainedInTheOtherCollection(){
+        addElementsToSet(set, st1, st2);
+        List<Student> list = Arrays.asList(st1, null, st2);
+
+        assertTrue(set.removeAll(list));
+        assertThat(set).doesNotContainAnyElementsOf(list);
+    }
+
+    @Test
+    void removeElementsShouldReturnNullPointerExceptionIfCollectionIsNull(){
+        Assertions.assertThrows(NullPointerException.class, () -> set.removeAll(null));
+    }
+
+    @Test
+    void shouldReturnTrueIfAllElementsFromOtherCollectionIsContainedInSetInAnyPosition(){
+        addElementsToSet(set, st1, st2, null);
+        List<Student> list = Arrays.asList(st1, null);
+
+        assertTrue(set.containsAll(list));
+    }
+
+    @Test
+    void retainElementsShouldReturnNullPointerExceptionIfCollectionIsNull(){
+        Assertions.assertThrows(NullPointerException.class, () -> set.retainAll(null));
+    }
+
+    @Test
+    void mustRetainAllElementsContainedInTheOtherCollection(){
+        addElementsToSet(set, st1, st2, null);
+        List<Student> list = Arrays.asList(st1, null);
+
+        set.retainAll(list);
+        assertThat(set).containsOnlyElementsOf(list);
+    }
+
+    @Test
+    void ifTheArrayIsLargerThanTheCollectionItShouldReturnAnArrayOfTheObjectWithTheSameSize(){
+        addElementsToSet(set, st1, st2, null);
+        Student[] students = set.toArray(new Student[5]);
+
+        assertAll(
+                () -> assertThat(students).hasSize(5),
+                () -> assertThat(students).containsAll(set)
+        );
+    }
+
+    @Test
+    void ifTheArrayIsSmallerThanTheCollectionItShouldReturnAnArrayOfTheObjectWithTheSameSizeAsSet(){
+        addElementsToSet(set, st1, st2, null);
+        Student[] students = set.toArray(new Student[1]);
+
+        assertAll(
+                () -> assertThat(students).hasSize(3),
+                () -> assertThat(students).containsAll(set)
+        );
+    }
+
+    @Test
+    void elementsInTheArrayShouldBeInTheSameOrderAsInBuckets(){
+        addElementsToSet(set, null, st1, st2);
+        Object[] students = set.toArray();
+        Object[] path = new Object[]{null, st2, st1};
+
+        assertArrayEquals(students, path);
+    }
+
+    @Test
+    void setShouldBeResizedIfSizeIsLargerThenLoadFactor(){
+        set = new StudentSet(1);
+        addElementsToSet(set, st1, st2, st3);
+
+        assertThat(set).hasSize(3);
+    }
+
+
+    private void addElementsToSet(StudentSet set, final Student ... students){
+        for(Student s : students){
+            assertTrue(set.add(s));
```

Comment:
this is a helper method, please remove this assert

Reply:
If I need it to be cleared or created new set after each test?

Reply:
Assert was removed for this method